### ISSUE TITLE
Expose accrued loan interest via listaccounthistory

### ISSUE THREAD

**Issue by RomanShumkov** [DeFiCh/ain#957]
#### What would you like to be added:
Expose accrued loan interest via listaccounthistory similar to liquidity mining rewards:
1. Accrued loan interest for each block and for each ticker symbol
2. New boolean parameter for listaccounthistory to show/hide interest entries (similar to no_rewards for liquidity mining rewards)

#### Why is this needed:
This is required for accounting and tax compliance and is blocking defichain friendly businesses from creating proper tax report for state authorities.

**Comment by bvbfan** [comment]
It will be a lot of useless data, we could do it via `getvaultinterest` ?

**Comment by RomanShumkov** [comment]
@bvbfan, I agree that it creates significant amount of data to store on the node for each new block, however I'm not sure if it's possible to have any kind of workaround without having more tricky/complicated implementation.

My use cases are:
1. Calculate accrued loan interest by month in crypto
2. Calculate accrued loan interest by month in fiat. In order to do that, I need to know accrued interest by day. Calculation is similar to rewards calculation (daily crypto amount is multiplied by fiat exchange rate of the particular day).

For my use cases it would be sufficient to have following set of data from the node:
<img width="452" alt="Screenshot 2021-12-07 at 17 17 04" src="https://user-images.githubusercontent.com/548534/145055981-8e6f4659-7d99-4309-a1e9-96579058a695.png">
API something like: `listaccruedinteresthistory vault-id` with optional params like the following:
1. time zone (UTC, UTC+2, UTC-2 etc.)
2. token (DUSD, GME, TSLA, etc.)
3. maxBlockHeight - default = chaintip
4. limit - default 100

This way we could reduce amount of records from 2880/day to 24/day - one record for each time zone for particular token symbol.

A bit more tricky to implement, but results in 100x more efficient storage.

Actually the same strategy could be used for rewards storage 100x optimization, but that's another story so let's just keep that in mind and not discuss it here :)

**Comment by RomanShumkov** [comment]
One more thing. In case of negative interest rates accrued interest (negative) would effectively be a reward (income). So maybe it makes sense to just go with the same implementation as for rewards/commissions at the beginning to keep it simple and straightforward:

> Expose accrued loan interest via listaccounthistory similar to liquidity mining rewards:
> 1. Accrued loan interest for each block and for each ticker symbol
> 2. New boolean parameter for listaccounthistory to show/hide interest entries (similar to no_rewards for liquidity mining rewards)

cc @bvbfan @prasannavl

**Comment by shohamc1** [comment]
Available via `getvault` now.

**Comment by RomanShumkov** [comment]
> Available via `getvault` as of 2.11.1.

@shohamc1, with `getvault` it's impossible to know accrued loan interest during particular time frame, e.g. last month or last year.

Let's say I'm now preparing tax report for Q1 2022. There is no way to calculate accrued loan interest during Q1 2022 without reimplementing consensus logic in my accounting software (calculating interest based on blocks and loans same as it's done by the full node).

Accrued loan interest should be exposed similarly to rewards - for each block or for each hour (by block median time) if we want to optimize resource utilization a bit.

**Comment by shohamc1** [comment]
Would it be feasible to go back in time using `invalidateblock` and using the `getvault` ouput to calculate the difference?

**Comment by RomanShumkov** [comment]
> Would it be feasible to go back in time using `invalidateblock` and using the `getvault` ouput to calculate the difference?

I had tried that several months ago, but it requires commenting out hardcoded checkpoints. And going back like 1 year ago triggers various consensus errors while "connecting" blocks which result in valid blocks being blacklisted by the node and requires even more source code tweaking on the go since full node was not designed/tested for this use case.

If we could make `invalidateblock` work perfectly with any depth it would be a decent workaround. It also enabled querying a lot of historical chain state by using existing rpc commands, which would run on specific block height instead of latest block.

One drawback is that it would take significant amount of time to sync chain to specific block in past (like 1 or 2 years ago), but if it's not failing and the use case is part of CI (i.e. stable and tested on regular basis) I would be happy to have it at least as a workaround :)

I would suggest invaludating your local chain to January 1st 2022 if you want to see specific errors node is complaining about when doing so.

Let me know if I can help with anything :)
